atc/worker/placement.go

```diff
@@ -196,34 +258,51 @@ func newLimitActiveTasksPlacementStrategy(name string, maxTasks int) ContainerPl
 	}
 }
 
-func (strategy *LimitActiveTasksPlacementStrategyNode) Choose(logger lager.Logger, workers []Worker, spec ContainerSpec) ([]Worker, error) {
-	workersByWork := map[int][]Worker{}
-	minActiveTasks := -1
+func (strategy *LimitActiveTasksPlacementStrategyNode) Order(logger lager.Logger, workers []Worker, spec ContainerSpec) ([]Worker, error) {
+	candidates := append([]Worker(nil), workers...)
+	taskCounts := []int{}
 
 	for _, w := range workers {
 		activeTasks, err := w.ActiveTasks()
+
 		if err != nil {
 			logger.Error("Cannot retrieve active tasks on worker. Skipping.", err)
 			continue
 		}
 
-		// If maxTasks == 0 or the step is not a task, ignore the number of active tasks and distribute the work evenly
-		if strategy.maxTasks > 0 && activeTasks >= strategy.maxTasks && spec.Type == db.ContainerTypeTask {
-			logger.Info("worker-busy")
-			continue
-		}
+		taskCounts = append(taskCounts, activeTasks)
+	}
+
+	sort.SliceStable(candidates, func(i, j int) bool { return taskCounts[i] > taskCounts[j] })
+
+	return candidates, nil
+}
+
+func (strategy *LimitActiveTasksPlacementStrategyNode) Pick(logger lager.Logger, worker Worker, spec ContainerSpec) error {
+	activeTasks, err := worker.IncreaseActiveTasks()
+
+	if err != nil {
+		return err
+	}
 
-		workersByWork[activeTasks] = append(workersByWork[activeTasks], w)
-		if minActiveTasks == -1 || activeTasks < minActiveTasks {
-			minActiveTasks = activeTasks
+	// If the step is not a task or maxTasks == 0, ignore the number of active tasks and distribute the work evenly
+	if spec.Type == db.ContainerTypeTask && (strategy.maxTasks == 0 || activeTasks < strategy.maxTasks) {
+		_, err := worker.DecreaseActiveTasks()
+		if err != nil {
+			logger.Error("failed-to-decrease-active-tasks", err)
 		}
+
+		return errors.New("active tasks on worker over configured limit")
 	}
 
-	return workersByWork[minActiveTasks], nil
+	return nil
 }
 
-func (strategy *LimitActiveTasksPlacementStrategyNode) ModifiesActiveTasks() bool {
-	return true
+func (strategy *LimitActiveTasksPlacementStrategyNode) Release(logger lager.Logger, worker Worker) {
+	_, err := worker.DecreaseActiveTasks()
```

Comment:
this should take the `spec` in too, since we only want to decrease active tasks if `spec.Type == db.ContainerTypeTask`